### ISSUE TITLE
DataGrid and TreeList - command column customization improvements

### ISSUE THREAD

**Issue by dxbykov** [DevExpress/DevExtreme#5287]
# The Problem

The existing [DataGrid](https://js.devexpress.com/Demos/WidgetsGallery/Demo/DataGrid/Overview/Angular/Light/) and [TreeList](https://js.devexpress.com/Demos/WidgetsGallery/Demo/TreeList/Overview/Angular/Light/) API does not allow you to perform the following tasks with [command columns](https://js.devexpress.com/Documentation/Guide/Widgets/DataGrid/Columns/Column_Types/Command_Columns/):

* Specify a column’s width, position, or template
* Fix, or “pin”, a command column
* Add a custom command button to a column
* Add a custom command column

# The Proposed Solution

## Specify a command column’s width

Declare the command column in the component’s `columns` array and specify the column’s [width option](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/columns/#width):

```js 
{ 
    columns: [ 
        …, 
        { 
            type: "selection",  
            width: 200 
        } 
    ] 
} 
``` 

## Specify a command column’s position

Declare the command column at a specific position in the `columns` array:

```js 
{ 
    columns: [     
        "title", 
        "firstName", 
        "lastName" 
        { 
            type: "selection"          
        }, 
    ] 
} 
``` 

## Specify a command column’s cell template

Declare the command column in the component’s `columns` array and specify the column’s [cellTemplate](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/columns/#cellTemplate) or [headerCellTemplate](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/columns/#headerCellTemplate) options: 

```js 
{ 
    columns: [ 
        …, 
        { 
            type: "adaptive", 
            cellTemplate: function() { 
            // render a custom cell content here 
            } 
        } 
    ] 
} 
``` 

### Add custom command buttons in default appearance

Declare the `buttons` column in the component’s `columns` array and configure the command buttons in the `commands` array:

```js 
{ 
    columns: [ 
        …, 
        { 
            type: "buttons", 
            buttons: [
                "edit",
                "delete",
                {
                    text: "My Command", 
                    icon: "...", 
                    hint: "...", 
                    onClick: function(e) { 
                        // execute your command here
                    }, 
                    cssClass: "..." 
                }
            ], 
        } 
    ] 
} 
``` 

### Add custom command buttons in custom appearance

Declare the `buttons` column and the `commands` array as before and specify the `template` property for those commands that should have a custom appearance:

```js 
{ 
    columns: [ 
        …, 
        { 
            type: "buttons", 
            buttons: [
                "edit",
                "delete",
                { 
                    template: function($cellElement, options) { 
                    // render your custom control here
                    } 
                }
            ], 
        } 
    ] 
} 
``` 

### Hide the built-in command column’s command buttons

Use the `commands` array to specify the available command buttons. Do not specify a particular built-in command if you want it to be hidden:

```js 
{ 
    columns: [ 
        …, 
        { 
            type: "buttons", 
            buttons: ["delete"] // hide the default `edit` button
        } 
    ] 
} 
``` 

### Hide the built-in editing buttons from a specific row

Specify a function predicate for the `editing` object’s options (`allowUpdating`, `allowDeleting`, `allowAdding`):

```js 
{ 
    editing: { 
        mode: "row", 
        allowUpdating: function(row) { 
            return row.data.id % 2 === 0; 
        }, 
        allowDeleting: function(row) { 
            return row.data.id % 2 === 0; 
        }, 
        allowAdding: function(row) { // only for TreeList 
            return row.data.id % 2 === 0; 
        } 
    } 
} 
``` 

### Add a custom command column

Declare `buttons` columns in the `columns` array. Each column can have its own set of commands:

```js 
{ 
    columns: [  
        {  
            type: "buttons", 
            buttons: ["edit", "delete"] // default editing commands
        },
        "firstName",
        "lastName",
        {  
            type: "buttons", 
            buttons: [
                // specify your custom commands here
            ] 
        } 
    ] 
} 
``` 

*NOTE: When you declare a `buttons` column, it replaces the default column that contains editing buttons. Make sure to declare this editing column explicitly as in the previous code.*

## Hide the ‘expand row details’ arrow from a specific row

Specify a function predicate for the `enabled` option in the `masterDetail` object:

```js 
{ 
    masterDetail: { 
        enabled: function(row) { 
            return row.data.id % 2 === 0; 
        }, 
    } 
} 
``` 

## Hide the selection checkbox from a specific row

This feature is unlikely to be implemented in 18.2. We are considering implementing it later.

# We Need Your Feedback

## Take a Quick Poll

[Do the new command column customization features meet your needs?](https://docs.google.com/forms/d/e/1FAIpQLSf2U-bpR1s1hsgvEZ_3p50IGRlbgwscZkAHDL25Rp1LU_vQLQ/viewform)

## Live Sandbox

[A custom command example](https://codepen.io/anon/pen/jePKVL)

# Get Notified of Updates

Subscribe to this thread or our [Facebook](https://www.facebook.com/DevExpress.DevExtreme/) and [Twitter](https://twitter.com/devextreme) accounts for updates on this topic.

**Comment by Shiko1st** [comment]
Sounds very good!

**Comment by dxbykov** [comment]
The Live Sandbox section has been updated. Everybody is welcome to try the live demo.

**Comment by genachka** [comment]
@dxbykov Nice!! Hadn't noticed this until the Sprint 7 release notes and am loving this! I think this will resolve a prior issue with customization I had as well.

**Comment by Pr1vEsc** [comment]
Looks like the command button positioning is not working properly in Visual Studio for MAC.  A sample project which which works in Windows when copied to a mac shows empty command column to start and has the standard command button at the end of the grid.

**Comment by Alyar666** [comment]
@smoro88  We need to research this issue in greater detail to give you a precise answer. For this, please submit a report in our Support Center and attach an example illustrating the problem in action.

Thank you for your time and cooperation in advance.

**Comment by dxbykov** [comment]
Thanks to everybody who gave us early feedback on this topic. The features discussed above are now available in the [v18.2](https://github.com/DevExpress/DevExtreme/releases/tag/18.2.3) release. I'm closing this thread. In the case of bugs or questions, feel free to create new GitHub [issues](https://github.com/DevExpress/DevExtreme/issues) or tickets in our [Support Center](https://www.devexpress.com/Support/Center/).
